Use AST transformations in `@tailwindcss/postcss`

RobinMalfait opened the pull request:
This PR improves the `@tailwindcss/postcss` integration by using direct AST transformations between our own AST and PostCSS's AST. This allows us to skip a step where we convert our AST into a string, then parse it back into a PostCSS AST.

The only downside is that we still have to print the AST into a string if we want to optimize the CSS using Lightning CSS. Luckily this only happens in production (`NODE_ENV=production`).

This also introduces a new private `compileAst` API, that allows us to accept an AST as the input. This allows us to skip the PostCSS AST -> string -> parse into our own AST step.

To summarize:

Instead of:
- Input: `PostCSS AST` -> `.toString()` -> `CSS.parse(…)` -> `Tailwind CSS AST`
- Output: `Tailwind CSS AST` -> `toCSS(ast)` -> `postcss.parse(…)` -> `PostCSS AST`

We will now do this instead:
- Input: `PostCSS AST` -> `transform(…)` -> `Tailwind CSS AST`
- Output: `Tailwind CSS AST` -> `transform(…)` -> `PostCSS AST`


---

Running this on Catalyst, the time spent in the `@tailwindcss/postcss` looks like this:
- Before: median time per run: 19.407687 ms
- After: median time per run: 11.8796455 ms

This is tested on Catalyst which roughly generates ~208kb worth of CSS in dev mode.

While it's not a lot, skipping the stringification and parsing seems to improve this step by ~40%.

Note: these times exclude scanning the actual candidates and only time the work needed for parsing/stringifying the CSS from and into ASTs. The actual numbers are a bit higher because of the Oxide scanner reading files from disk. But since that part is going to be there no matter what, it's not fair to include it in this benchmark.

RobinMalfait committed:
allow to pass in the `!important` flag

RobinMalfait committed:
implement `postCssAstToCssAst` and `cssAstToPostCssAst`

These methods can convert PostCSS ASTs to our internal CSS ASTs and vice
versa. This allows us to skip a step where introduce
parsing/stringification.

Instead of:
- `PostCSS AST` -> `.toString()` -> `postcss.parse` -> `CSS AST`
- `CSS AST` -> `toCSS(ast)` -> `CSS.parse` -> `PostCSS AST`

We will now do this instead:
- `PostCSS AST` -> `transform(…)` -> `CSS AST`
- `CSS AST` -> `transform(…)` -> `PostCSS AST`

RobinMalfait committed:
optimize AST before printing

This introduces an `optimizeAst(…)` function that creates a fresh AST
from an AST but handles all the edge cases we used to have in
`toCss(…)`.

For example, `@property` is deduped (and fallbacks are generated for
older browsers), `Context` nodes are transparent and `AtRoots` move
nodes into the actual root.

This allows us to simplify the `toCss(…)` code to be a 1-to-1
translation and simply print `declarations`, `rules`, `at-rules` and
comments. We don't have to worry about the other special logic.

This also means that we don't have to re-introduce the same logic in our
PostCSS AST <-> Our AST transformations.

---

In addition: I ran some checks on the Catalyst codebase, and noticed
that we have the most declarations, then rules, then at-rules, then
at-roots, then context nodes and last but not least comments.

With this information in mind, the if-branches are using this
information to first check for declarations, then rules, ...

RobinMalfait committed:
implement new `compileAst(…)` and update `compile(…)`

RobinMalfait committed:
expose `compileAst` from `@tailwindcss/node`

RobinMalfait committed:
WIP: PostCSS implementation

thecrypticace committed:
Tweak code

thecrypticace committed:
Convert our AST to PostCSS’s AST

thecrypticace committed:
Pass through source information

RobinMalfait committed:
print CSS from our own AST

RobinMalfait committed:
only convert our AST into a PostCSS AST when not optimizing

If we are not optimizing, we don't need the `context.cachedAst` because
we will use the optimziedAst instead.

RobinMalfait committed:
remove unused variable

philipp-spiess commented on packages/@tailwindcss-node/src/compile.ts:
```diff
@@ -7,15 +7,80 @@ import { pathToFileURL } from 'node:url'
 
 export type Resolver = (id: string, base: string) => Promise<string | false | undefined>
 
+export async function compileAst(
```
is it possible we share some code here between the two implementations?

philipp-spiess commented on packages/tailwindcss/src/css-parser.ts:
```diff
@@ -434,15 +435,7 @@ export function parse(input: string) {
-                .trim(),
-              important: importantIdx !== -1,
-            } satisfies Declaration)
+            parent.nodes.push(parseDeclaration(buffer, colonIdx))
```
nice cleanup

philipp-spiess commented on packages/tailwindcss/src/index.ts:
```diff
@@ -580,7 +581,7 @@ export async function compile(
   // and should be ignored.
   let allValidCandidates = new Set<string>()
-  let compiledCss = features !== Features.None ? toCss(ast) : css
+  let compiled = features !== Features.None ? optimizeAst(ast) : input
```
I think we can actually get rid of this `optimizeAst()` call here. There are two scenarios to consider:

1. A subsequent `build()` is called with no changes: This is the case where we currently safe one computation. In that case we can also move the `optimizeAst()` inot the `build()` function though and add something like (`if (!compiled) features !== Features.None ? optimizeAst(ast) : input`) there inline
2. A subsequent `build()` is called **with changes**: I'd say this happens very often since the first build probably more often than not does require candidates which will require a rebuild. In this case we're currently computing the initial `optimizeAst` pass for no reason since it's going to be replaced soon anyway.

This is possible since the `compileAst()` function does never return the compiled CSS, you can only do so by calling `build()`.

philipp-spiess commented on packages/tailwindcss/src/ast.ts:
```diff
@@ -208,18 +208,151 @@ export function walkDepth(
-  let atRoots: string = ''
+// Optimize the AST for printing where all the special nodes that require custom
+// handling are handled such that the printing is a 1-to-1 transformation.
+export function optimizeAst(ast: AstNode[]) {
```
Hm now that I see this in the whole PR I'm not so super convinced about this name anymore. The reason being that we refer to `optimize` as the step in post-compilation where we run `lightningcss`. Now we're overloading the term with something used during the compilation. Maybe `applyFallbacksToAst` or something like this? 🤔 I also dig your initial name of `prepareAstForPrinting` honestly lol. Find it less confusing then overloading the term.

RobinMalfait replied:
Hmm I see. The problem is that this is doing multiple things. While `applyFallbacksToAst` is correct, we are also in addition to tracking the fallbacks making sure that `@property` is handled once, that `Context` nodes become transparent and that `AtRoot` nodes are hoisted.

`prepareAstForPrinting` is a bit more vague which is good to describe that we are doing multiple things. But the `forPrinting` part is off because in `@tailwindcss/postcss` that would be `prepareAstForTransformingIntoOtherAST` 😅

philipp-spiess commented on packages/@tailwindcss-postcss/src/index.ts:
```diff
@@ -205,23 +211,29 @@ function tailwindcss(opts: PluginOptions = {}): AcceptedPlugin {
+              env.DEBUG && console.timeEnd('[@tailwindcss/postcss] Optimize CSS')
+            } else {
+              // Convert our AST to a PostCSS AST
+              context.cachedAst = cssAstToPostCssAst(ast, root.source)
```
should this be wrapped in a timing call as well for debug cases?

RobinMalfait replied:
Yeah good catch!

RobinMalfait committed:
rename variables to be more explicit

So many ASTs, but which ones?

RobinMalfait committed:
add timing info around `cssAstToPostCssAst(…)`

RobinMalfait committed:
restructure `build(…)`

This allows us to bail earlier and delay the `optimizeAst(…)` call and
only do it when it's necessary.

RobinMalfait committed:
track the `tailwindCssAst` for subsequent cache hits

RobinMalfait committed:
ensure PostCSS AST nodes have semicolons

RobinMalfait committed:
handle comments correctly

1. If a PostCSS comment is coming in, only map it to our AST's comment
   node if it starts with `!` (license comments).
2. If a comment from our AST is mapped to a PostCSS AST Comment, then
   remove the default whitespace and rely on the whitespace encoded in
   `node.value`.

RobinMalfait committed:
compute current_mtimes in parallel per file/directory

RobinMalfait committed:
remove instrumentation from `read_dir`

RobinMalfait committed:
trick PostCSS into using 2 spaces instead of 4

Bonus points: this also means that PostCSS doesn't need to do a walk to
try and detect the indentation.

RobinMalfait committed:
use `??=` shorthand

RobinMalfait committed:
ensure `compile` and `compileAst` re-use the same logic

RobinMalfait committed:
run format

Unrelated to this PR, but noticed it when running `prettier`

RobinMalfait committed:
clone the cached PostCSS tree before appending it to the root

philipp-spiess reviewed (approved):
Superb stuff
